node/core/approval-voting/src/ops.rs

```diff
@@ -0,0 +1,379 @@
+			Some(_) => None,
+		};
+
+		new_range.map(|n| store.write_stored_block_range(n));
+	};
+
+	// Update the blocks at height meta key.
+	{
+		let mut blocks_at_height = store.load_blocks_at_height(&number)?;
+		if blocks_at_height.contains(&entry.block_hash()) {
+			// seems we already have a block entry for this block. nothing to do here.
+			return Ok(Vec::new())
+		}
+
+		blocks_at_height.push(entry.block_hash());
+		store.write_blocks_at_height(number, blocks_at_height)
+	};
+
+	let mut candidate_entries = Vec::with_capacity(entry.candidates().len());
+
+	// read and write all updated entries.
+	{
+		for &(_, ref candidate_hash) in entry.candidates() {
+			let NewCandidateInfo {
+				candidate,
+				backing_group,
+				our_assignment,
+			} = match candidate_info(candidate_hash) {
+				None => return Ok(Vec::new()),
+				Some(info) => info,
+			};
+
+			let mut candidate_entry = store.load_candidate_entry(&candidate_hash)?
+				.unwrap_or_else(move || CandidateEntry {
+					candidate,
+					session,
+					block_assignments: BTreeMap::new(),
+					approvals: bitvec::bitvec![BitOrderLsb0, u8; 0; n_validators],
+				});
+
+			candidate_entry.block_assignments.insert(
+				entry.block_hash(),
+				ApprovalEntry {
+					tranches: Vec::new(),
+					backing_group,
+					our_assignment: our_assignment.map(|v| v.into()),
+					our_approval_sig: None,
+					assignments: bitvec::bitvec![BitOrderLsb0, u8; 0; n_validators],
+					approved: false,
+				}
+			);
+
+			store.write_candidate_entry(candidate_entry.clone());
+
+			candidate_entries.push((*candidate_hash, candidate_entry));
+		}
+	};
+
+	// Update the child index for the parent.
+	store.load_block_entry(&parent_hash)?.map(|mut e| {
+		e.children.push(entry.block_hash());
+		store.write_block_entry(e);
+	});
+
+	// Put the new block entry in.
+	store.write_block_entry(entry);
+
+	Ok(candidate_entries)
+}
+/// Forcibly approve all candidates included at up to the given relay-chain height in the indicated
+/// chain.
+pub(super) fn force_approve(
+	store: &mut OverlayedBackend<'_, impl Backend>,
+	chain_head: Hash,
+	up_to: BlockNumber,
+) -> SubsystemResult<()> {
+	enum State {
+		WalkTo,
+		Approving,
+	}
+
+	let mut cur_hash = chain_head;
+	let mut state = State::WalkTo;
+
+	// iterate back to the `up_to` block, and then iterate backwards until all blocks
+	// are updated.
+	while let Some(mut entry) = store.load_block_entry(&cur_hash)? {
+
+		if entry.block_number() <= up_to {
+			state = State::Approving;
+		}
+
+		cur_hash = entry.parent_hash();
+
+		match state {
+			State::WalkTo => {},
+			State::Approving => {
+				entry.approved_bitfield.iter_mut().for_each(|mut b| *b = true);
+				store.write_block_entry(entry);
+			}
+		}
+	}
+
+	Ok(())
+}
+/// Return all blocks which have entries in the DB, ascending, by height.
+pub fn load_all_blocks(store: &dyn KeyValueDB, config: &Config) -> SubsystemResult<Vec<Hash>> {
```

Comment:
This stuff seems a little out of place in `ops.rs` because it accesses the `KeyValueDB` (concrete schema) directly. The rest of this file is high-level operations around the `Backend` abstraction.